cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/DefaultDatasetRuntimeContext.java

```diff
@@ -0,0 +1,222 @@
+
+  // Just use simple primitive to memorize whether a particular access type has lineage recorded or not
+  // Don't worry about concurrency because
+  // 1. doesn't matter too much if the lineage get recorded again from different thread, but it'll be only once
+  //    per type per thread
+  // 2. the lineage writer implementation anyway has a concurrent map for caching. The check here acts as a
+  //    low cost gate instead of hitting the ConcurrentMap.putIfAbsent method with new object creation on
+  //    every DS operation call.
+  private boolean recordedNoAccess;
+  private boolean recordedRead;
+  private boolean recordedWrite;
+  private boolean recordedReadWrite;
+
+  /**
+   * Helper method to execute a {@link Callable} with a {@link DatasetRuntimeContext}.
+   * This method is mainly called from
+   * {@link LineageWriterDatasetFramework#getDataset(Id.DatasetInstance, Map, ClassLoader, DatasetClassLoaderProvider,
+   * Iterable, AccessType)}.
+   */
+  public static <T> T execute(AuthorizationEnforcer enforcer,
+                              DatasetAccessRecorder accessRecorder,
+                              Principal principal,
+                              DatasetId datasetId,
+                              Callable<T> callable) throws Exception {
+    // Memorize the old context, change to a new one and restore it at the end.
+    // It is needed so that nested call to DatasetFramework.getDataset can create the call site context correctly.
+    Cancellable cancel = setContext(new DefaultDatasetRuntimeContext(enforcer, accessRecorder, principal, datasetId));
+    try {
+      return callable.call();
+    } finally {
+      cancel.cancel();
+    }
+  }
+
+  private DefaultDatasetRuntimeContext(AuthorizationEnforcer enforcer, DatasetAccessRecorder accessRecorder,
+                                       Principal principal, DatasetId datasetId) {
+    this.enforcer = enforcer;
+    this.accessRecorder = accessRecorder;
+    this.principal = principal;
+    this.datasetId = datasetId;
+  }
+
+  @Override
+  public void onMethodEntry(@Nullable Class<? extends Annotation> annotation,
+                            Class<? extends Annotation> defaultAnnotation) {
+    final Deque<Class<? extends Annotation>> callStack = this.callStack.get();
+    Class<? extends Annotation> parentAnnotation = callStack.peekLast();
+
+    if (annotation == null) {
+      // If there is no annotation on the method, either use immediate parent one from the call stack
+      // or use the default one if the call stack is empty
+      annotation = (parentAnnotation == null) ? defaultAnnotation : parentAnnotation;
+    }
+
+    // Do a quick check if the current permission is allowed by the immediate parent if there is one.
+    if (parentAnnotation != null && !isAllowed(parentAnnotation, annotation)) {
+      throw new DataSetException(
+        String.format("Current scope only allows @%s, but the current method requires @%s",
```

Comment:
This logic seems correct. But: IndexedObjectStore.write() is annotated with @WriteOnly, and it calls Table.get() which is @ReadOnly. So that should trigger this exception. Either we don't have test cases for IndexedObjectStore, or there is a bug in the construction of the call stack.